### ISSUE TITLE
Injecting helper-objects into POJO-rules

### ISSUE THREAD

**Issue by pjotre86** [rulebook-rules/rulebook#111]
Hey!
I have one more question.
Imagine the case I need some utility-object (f.e. for db-access) in my POJO-rule. Is there any other way to inject this object into the rule than injecting it as a fact?
Because this feels kind of wrong to me, because the object is not really a fact which is needed to handle my business-rules, it's more a helper-object... I guess you know what I mean.
In a spring-context f.e. it would be cool if something like that would work (I mean the dao-field in the middle):
```
@Rule( name = "SimpleRule")
public class SimpleRule {
    @Given("status")
    private int status;
    @Given("object")
    private Object object;
    @Result
    private String result;

    @Autowired 
    private Dao dao;

    @When
    public boolean when() {
        return true;
    }

    @Then
    public RuleState then() {
        dao.iWantToLoadSomethingToFigureOutWhatToDoNow();
        result = "Rule was triggered with status=" + status + " and object=" + object;
        return  RuleState.NEXT;
    }
}
```
Do you have any suggestion how to do this by not injecting this as a Fact?
Regards,
Peter

**Comment by Clayton7510** [comment]
Right now, I think it would need to be fact. But I like where you are going with this. I think there would be some value in having rules be container aware for Spring and maybe Weld too.

The RuleBookRunner takes care of instantiating the POJO Rules and RuleAdapter takes care of adapting POJO rules to the Rule interface. So, you could modify RuleAdapter or create another Spring container-aware RuleBookRunner that did allow for @Autowire to work. I'd probably do the latter. 

Thanks for the suggestion. I'll add it to a future release. Or if you want to try to bite it off yourself, and submit a PR, that would be cool too.

**Comment by walesfang** [comment]
I should say i am in the similar context in my working project:
e.g. my RuleBook need to call other services (rpc) to get a list of POJO (templates) which is the skeleton of building my output result regarding to different business rules defined in my rule book.
Will also try have a look on this.

**Comment by pjotre86** [comment]
I'd be happy to help out. But before I'd start coding I would give it some more thoughts. I'm pretty sure that there is some value in having the possibility to inject utility-objects into rules, especially in a bit more complex applications I think it's is a very likely necessity.
But to keep the API of rulebook clean I would not just consider this in a spring (or weld, guice whatever) context. I would also make it possible for "plain" rulebook users, by adding an annotation like "@Utility" ore something, so you don't have to "abuse" facts for that. I think this is even more important than having is possible in Spring. Because otherwise you would make the framework (or at least some nice features of it) kind of dependend to Spring etc.
What's your opinion?

**Comment by Clayton7510** [comment]
@pjotre86, I like how you are thinking about not wanting to be tied into one specific container. But I also think there is value in using the syntax that developers are used to. There is also a JSR standard for injecting resources that Spring supports (not sure about Guice tho). But regardless, the implementation is going to be a little different for each provider.

I would say, start with Spring and allow for @Autowired and @Inject, @Inject being the standard that works outside of Spring. RuleBook will have to be the one actually facilitating the injection via a RuleBookRunner implementation, like RuleBookSpringRunner; it will be a facade on top of the spring container. And it will have to be bound to Spring. Why? Because POJO rule instances aren't created until RuleBookRunner is run. You could do this as a decorator, perhaps. Play with it and see what you can do that could be used in place of the standard RuleBookRunner without ricocheting across a bunch of RuleBook classes and done in a relatively simple fashion. 

You could do this either in rulebook-spring and take all the Spring stuff outside of RuleBook or you could do it inside of rulebook-core and then folks who aren't using Spring can just exclude the Spring dependencies (kind of a pain, but do-able). Either way, we'll move things around if we need to.

**Comment by pjotre86** [comment]
All right, sounds like a plan to me!
I have to say I'm not a spring-expert, but I will use this gain some experience/knowledge there. My schedule is abit tight, as I also have a job + a new girlfriend ;-) but I think latest on friday I'll find some time to look into it.

**Comment by Clayton7510** [comment]
I don't know many people who do this stuff as a full-time gig. It's just about learning, having fun and giving a little back as time allows. If you have questions, feel free to post on here or the gitter channel. And if you find that you don't have the time to allocate for completion, post that too and someone else can pick it up.

**Comment by ekoon** [comment]
Use SpringRuleBook and @RuleBean.  Create the rulebook as a SpringRuleBook and a new bean for your POJO.  Annotate your POJO with @RuleBean.  Your POJO now supports @Autowire.

**Comment by joannbrereton** [comment]
@ekoon Can you provide an example of the use of @rulebean ?

**Comment by lanegoolsby** [comment]
@joannbrereton, the examples for using @Rule and @RuleBean do not seem to work with Spring Boot apps. However, if you annotate your rule with _both_ annotations things seem to work as shown in the examples.

**Comment by lanegoolsby** [comment]
I was able to get dependencies autowired into the rules by adding both Rule and RuleBean annotations to the rule class. It appears that RuleBean does not implement the Rule interface so the reflection query does not pick them up.

**Comment by Panthro** [comment]
What's the stage of this? I'm currently trying to get my rules to get @components injected and there's no way.

Is there an example somewhere to get this to work?

thanks

**Comment by Clayton7510** [comment]
No PRs for this fix. I am cleaning up a few outstanding bugs at the moment. Once that's done, the focus will be Spring enhancements, including this feature.

**Comment by Clayton7510** [comment]
@Panthro actually, there is an update in v0.11-SNAPSHOT that allows components to be @Autowired inside of Rules. Check out the com.deliveredtechnologies.rulebook.spring.SpringAwareRuleBookRunnerTest. I think that might be what you are looking for. There is currently nothing available for Weld, however.

**Comment by Clayton7510** [comment]
I put an very trivial example here: https://github.com/rulebook-rules/rulebook#5-using-rulebook-with-spring. But it demonstrates the use.

**Comment by Clayton7510** [comment]
This was a pretty recent contribution from @rizroj. So, it's only in the 0.11-SNAPSHOT until the next release (within the next week). Up until that contribution, rulebook- spring didn't really offer much beyond a few things for convenience. Now, with SpringAwareRuleBookRunner, rulebook- spring is needed to have Spring injectable rules. Sorry for the confusion. I updated the Spring section in the readme on the develop branch with a trivial example that shows how to have Spring inject components into rules.
